### ISSUE TITLE
[BUG] Potential bugs on vm-dhcp-controller

### ISSUE THREAD

**Issue by w13915984028** [harvester/harvester#5072]
**Describe the bug**

Potential bugs on vm-dhcp-controller, please double check and optimze in future releases.

1. The `EventListener` blocks on `select {}`, but `(c *Controller) Run` may fail when ` !cache.WaitForCacheSync(stopCh, c.informer.HasSynced)`; if it happens, it blocks

2. The `context` in `(a *Agent) Run()`, is not passed to later procssing, the program can only be terminated forcely without early termination.

```
func (e *EventHandler) EventListener() (err error) {
...
	controller := NewController(queue, indexer, informer, e.poolRef, e.dhcpAllocator, e.poolCache)
	stop := make(chan struct{})
	defer close(stop)
	go controller.Run(1, stop)

	select {}
}

func (c *Controller) Run(workers int, stopCh chan struct{}) {
	defer runtime.HandleCrash()

	defer c.queue.ShutDown()
	logrus.Infof("(ippool.Run) starting IPPool controller")

	go c.informer.Run(stopCh)
	if !cache.WaitForCacheSync(stopCh, c.informer.HasSynced) {
		logrus.Errorf("(ippool.Run) timed out waiting for caches to sync")

		return
	}

	for i := 0; i < workers; i++ {
		go wait.Until(c.runWorker, time.Second, stopCh)
	}

	<-stopCh
	logrus.Infof("(ippool.Run) stopping IPPool controller")
}

func (a *Agent) Run() error {
	logrus.Infof("monitor ippool %s", a.poolRef.String())

	eg, egctx := errgroup.WithContext(a.ctx)

..
	eg.Go(func() error {
		select {
		case <-egctx.Done():
			return nil
		default:
			// initialize the ippoolEventListener handler
			if err := a.ippoolEventHandler.Init(); err != nil {
				logrus.Fatal(err)
			}
			return a.ippoolEventHandler.EventListener()
		}
	})

	if err := eg.Wait(); err != nil {
		logrus.Fatal(err)
	}

	return nil
}
```
3. The `name := fmt.Sprintf("%s-%s-agent", ipPool.Namespace, ipPool.Name)` may over 63 chars. (discussed before)

4. The `ippool-register` will try to create a POD at any change of the ip-pool status, should `status.AgentPodRef` be helpful to check if the POD exists first, avoid un-necessary API call to create and get an `IsAlreadyExists` error.

If the agent POD's namespace & name may change (not confirmed), then the `status.AgentPodRef` can also be used to delete previous POD and re-create the new one.

```
	ctlnetworkv1.RegisterIPPoolStatusHandler(
		ctx,
		ippools,
		networkv1.Registered,
		"ippool-register",
		handler.DeployAgent,
	)

func (h *Handler) DeployAgent(ipPool *networkv1.IPPool, status networkv1.IPPoolStatus) (networkv1.IPPoolStatus, error) {
	agent := prepareAgentPod(ipPool, h.noDHCP, h.agentNamespace, clusterNetwork, h.agentServiceAccountName, h.agentImage)

	agentPod, err := h.podClient.Create(agent)
	if err != nil {
		if apierrors.IsAlreadyExists(err) {
			return status, nil
		}

	status.AgentPodRef = &networkv1.PodReference{
		Namespace: agentPod.Namespace,
		Name:      agentPod.Name,
	}		
```

5. The sequence question: the `BuildCache` and `Allocate` are running in parallel, and they are not using a high level control like `networkv1.CacheReady`

```
func (h *Handler) BuildCache(ipPool *networkv1.IPPool, status networkv1.IPPoolStatus) (networkv1.IPPoolStatus, error) {
...
h.cacheAllocator.AddMAC(ipPool.Spec.NetworkName, mac, ip)
...

func (h *Handler) Allocate(vmNetCfg *networkv1.VirtualMachineNetworkConfig, status networkv1.VirtualMachineNetworkConfigStatus) (networkv1.VirtualMachineNetworkConfigStatus, error) {
...
exists, err := h.cacheAllocator.HasMAC(nc.NetworkName, nc.MACAddress)
...
		// Recover IP from status (resume from paused state)
		if oIP, err := findIPAddressFromNetworkConfigStatusByMACAddress(vmNetCfg.Status.NetworkConfig, nc.MACAddress); err == nil {
			dIP = oIP
		}

		ip, err := h.ipAllocator.AllocateIP(nc.NetworkName, dIP)
		if err != nil {
			return status, err
		}

		if err := h.cacheAllocator.AddMAC(nc.NetworkName, nc.MACAddress, ip); err != nil {
			return status, err
		}
...
```

**To Reproduce**

**Expected behavior**

**Support bundle**

**Environment**
 - Harvester ISO version:  v1.3.0 master-head
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630):

**Additional context**

Observed when reviewing PR: https://github.com/harvester/harvester/pull/4960

**Comment by w13915984028** [comment]
Issue 6:

Another question is about those 2 local cached data, it is somewhat anti k8s-controller patten. In short, the k8s controllers are looping CRUD on CRD objects.

Now, the controller will build those 2 in-memory cache when controller starts, and keep syncing them with CRD objects. They are kind of `view` of `IPPool` and `VMNetworkConfig` CRD objects.

As mentioned in issue 5, this increases the chance of data race, and data consistency will be a topic.

Are there some reasons that those 2 are required instead of `CRUD` from the CRD objects?

https://github.com/harvester/vm-dhcp-controller/blob/ef900c6bcf6e16b31edec39152383aea4524867f/pkg/ipam/ipam.go#L13
```
type IPSubnet struct {
	ipNet     *net.IPNet
	start     net.IP
	end       net.IP
	broadcast net.IP
	ips       map[string]bool
}

type IPAllocator struct {
	ipam  map[string]IPSubnet
	mutex sync.RWMutex
}
```


https://github.com/harvester/vm-dhcp-controller/blob/ef900c6bcf6e16b31edec39152383aea4524867f/pkg/cache/cache.go#L9
```
type MACSet struct {
	macs map[string]net.IP
}

type CacheAllocator struct {
	cache map[string]MACSet
	mutex sync.RWMutex
}
```

**Comment by lanfon72** [comment]
@starbops please add milestone, thanks.

**Comment by irishgordo** [comment]
**Validation Issue:**
@starbops 

From, https://github.com/harvester/vm-dhcp-controller/pull/19#issue-2114780175 , for:

```yaml

╭─mike at suse-workstation-team-harvester in ~/Projects/test-5065/vm-dhcp-controller on main✔
╰─± kubectl get network-attachment-definitions -A
NAMESPACE   NAME                                                              AGE
default     dhcp900                                                           3h22m
default     fi6cx9ca1kt1faq80k3ro9cowyumyjb67qdmg8fb9ydmz27rbk5btlg2m5avv3n   2m41s
default     test                                                              6d1h
default     test-net                                                          22h
default     vlan104                                                           3d21h
╭─mike at suse-workstation-team-harvester in ~/Projects/test-5065/vm-dhcp-controller on main✔
╰─± cat ../ippool.yaml 
apiVersion: network.harvesterhci.io/v1alpha1
kind: IPPool
metadata:
  namespace: default
  name: fi6cx9ca1kt1faq80k3ro9cowyumyjb67qdmg8fb9ydmz27rbk5btlg2m5avv3n
spec:
  ipv4Config:
    serverIP: 192.168.30.200
    cidr: 192.168.30.1/24
    pool:
      start: 192.168.30.2
      end: 192.168.30.250
    router: 192.168.30.1
    leaseTime: 300
  networkName: default/fi6cx9ca1kt1faq80k3ro9cowyumyjb67qdmg8fb9ydmz27rbk5btlg2m5avv3n
╭─mike at suse-workstation-team-harvester in ~/Projects/test-5065/vm-dhcp-controller on main✔
╰─± kubectl create -f ../ippool.yaml
The IPPool "fi6cx9ca1kt1faq80k3ro9cowyumyjb67qdmg8fb9ydmz27rbk5btlg2m5avv3n" is invalid: 
* spec.networkName: Too long: may not be longer than 64
* <nil>: Invalid value: "null": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation
╭─mike at suse-workstation-team-harvester in ~/Projects/test-5065/vm-dhcp-controller on main✔
╰─± kubectl describe network-attachment-definitions/fi6cx9ca1kt1faq80k3ro9cowyumyjb67qdmg8fb9ydmz27rbk5btlg2m5avv3n | yq
Error: bad file '-': yaml: line 7: mapping values are not allowed in this context
╭─mike at suse-workstation-team-harvester in ~/Projects/test-5065/vm-dhcp-controller on main✔
╰─± kubectl get network-attachment-definitions/fi6cx9ca1kt1faq80k3ro9cowyumyjb67qdmg8fb9ydmz27rbk5btlg2m5avv3n -o yaml | yq
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  annotations:
    network.harvesterhci.io/route: '{"mode":"auto","connectivity":"DHCP failed"}'
  creationTimestamp: "2024-03-05T23:05:06Z"
  finalizers:
    - wrangler.cattle.io/harvester-network-nad-controller
    - wrangler.cattle.io/harvester-network-manager-nad-controller
  generation: 1
  labels:
    network.harvesterhci.io/clusternetwork: cn-dhcptest
    network.harvesterhci.io/ready: "true"
    network.harvesterhci.io/type: L2VlanNetwork
    network.harvesterhci.io/vlan-id: "30"
  name: fi6cx9ca1kt1faq80k3ro9cowyumyjb67qdmg8fb9ydmz27rbk5btlg2m5avv3n
  namespace: default
  resourceVersion: "1672667"
  uid: 00d2713c-9fc0-48bc-8a0f-5c8630d0d418
spec:
  config: '{"cniVersion":"0.3.1","name":"fi6cx9ca1kt1faq80k3ro9cowyumyjb67qdmg8fb9ydmz27rbk5btlg2m5avv3n","type":"bridge","bridge":"cn-dhcptest-br","promiscMode":true,"vlan":30,"ipam":{}}'

```

It seems to not create even though the `spec.networkName` is valid via `kubectl get network-attachment-definitions -A`.

**Comment by irishgordo** [comment]
@starbops 
Still noticing this as an issue on:
- v1.3.0-rc4
- 0.3.0 chart version with v0.3.0 tags on the images

```
╭─mike at suse-workstation-team-harvester in ~/Projects/test-5065
╰─○ kubectl create -f ippool.yaml 
The IPPool "fi6cx9ca1kt1faq80k3ro9cowyumyjb67qdmg8fb9ydmz27rbk5btlg2m5avv3n" is invalid: 
* spec.networkName: Too long: may not be longer than 64
* <nil>: Invalid value: "null": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation

```

**Comment by starbops** [comment]
> @starbops Still noticing this as an issue on:
> 
> * v1.3.0-rc4
> * 0.3.0 chart version with v0.3.0 tags on the images
> 
> ```
> ╭─mike at suse-workstation-team-harvester in ~/Projects/test-5065
> ╰─○ kubectl create -f ippool.yaml 
> The IPPool "fi6cx9ca1kt1faq80k3ro9cowyumyjb67qdmg8fb9ydmz27rbk5btlg2m5avv3n" is invalid: 
> * spec.networkName: Too long: may not be longer than 64
> * <nil>: Invalid value: "null": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation
> ```

The bug fix harvester/vm-dhcp-controller#19 has been completely shadowed by the enhancement harvester/vm-dhcp-controller#18, so it becomes unnecessary to validate this one. I suggest we skip this one since #5065 has been closed. Now the `.spec.networkName` cannot longer than 64 characters.

Thanks!

cc @irishgordo

**Comment by irishgordo** [comment]
Sounds good @starbops  I'll go ahead and close this one out then :+1: :smile:
